### DISCUSSION TITLE
DRM lease manager

### DISCUSSION THREAD

**Discussion by garlett** [labwc/labwc#1198] [Ideas]
Is it possible to add a command line argument that uses a CRT lease ?

This is used to enable multiple seats on the same gpu. [Example](https://github.com/garlett/multiseat)

Automotive linux made "Drm Lease Manager" that splits the gpu into leases, that can be used on a patched version of weston. But weston is too much minimalist.

Or is this a feature better suited for wlroots ?

Thx

**Comment by Consolatis** [comment]
labwc does support DRM leasing. But as far as I understood the intended use-case it is designed so applications could take over complete control of a screen like for VR headsets.

Edit:
Oh I see, you want the client part of that protocol. That would indeed be something to ask the wlroots devs. Feel free to paste a link to your wlroots issue here though, I'd be interested to follow that discussion as well. (Small suggestion: don't mention anything about Nvidia GPUs there).

**Reply by garlett** [reply]
Ok, thank you, I will include the client part you mentioned. And yes, fk u NVIDIA !!!

**Comment by garlett** [comment]
https://gitlab.freedesktop.org/wlroots/wlroots/-/issues/3750

**Comment by garlett** [comment]
Segments copied from IRC:

**kennylevinsen**
hmm I wonder what happens when you lease out a connector on intel with regards to resource allocations during modeset...
I mean in the case where e.g. a lease manager has leased out to two independent display servers
each of the display servers will need to modeset independently, unable to include other crtcs in their commits

this would either be in the automotive case with the "drm lease manager" thing, or in the case of seatd handing out leases to subdivide a gpu among seats
... the drm-lease-manager sends *pointers* on the wire to dlm clients and blindly dereferences them, and here I thought my habit of sending structs was naughty


**Consolatis**
it might actually have nothing to do with the drm-lease wayland protocol: https://docs.automotivelinux.org/en/lamprey/5_Component_Documentation/4_drm-leasemanager/
and is instead referring to the kernel feature itself + in this case it seems they use some external daemon + library to manage drm lease fds, which seems to be this one: https://gerrit.automotivelinux.org/gerrit/gitweb?p=src/drm-lease-manager.git;a=blob;f=README.md

maybe a wlroots implementation in the drm backend to accept lease fds + making that function public to compositors and the wayland backend could potentially support both use cases? It could either be used when running nested (using the wayland drm lease protocol) or run based on a lease fd acquired by unspecified means by the compositor (like inheriting the fd during an exec of the compositor or by implementing whatever that "drm lease manager" protocol 

one may argue that the env vars like WLR_RENDER_DRM_DEVICE serve as somewhat of a CLI arg. Although.. It could be used for the acquire-lease-fd-and-exec-compositor case by having the fd int in the env var before executing the compositor.
 they could design a wrapper that uses that protocol (or other means to acquire a lease fd), put the fd int into an env var and then exec the compositor which then inherits the fd, and wlroots would then try to use that fd as a drm lease lease

**kennylevinsen**
 depending on the goal, you could patch seatd to hand over a leased fd instead of the real drm fd, I have been planning to look at such sub-divided seats, but never got around to it, then wlroots won't need to know.

**Comment by garlett** [comment]
I was going to modify the weston patch that support drm lease and insert on [wlr_session_find_gpus](https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/backend/session/session.c?ref_type=heads#L450), using an environment variable to pass the lease fd, but labwc does not have multiseat support

**Comment by Consolatis** [comment]
Labwc doesn't support multiple seats per instance running.
However, spawning a separate instance for each seat should be fine. Then it would be up to seatd to provide the correct set of inputs and drm fd to wlroots and thus labwc. There should be no changes required for either wlroots or labwc that way.

Note that the last time I personally looked into multiseat was before systemd was even a thing so take my response with a grain of salt.
I am also making the assumption here that wlroots gets all input devices and drm fds from seatd rather than discovering everything on its own.

**Comment by garlett** [comment]
I was talking with kennylevinsen about seatd support both things, 

Seatd would create the leases without drm lease manager from automotive Linux,  no code done.
The wlr_session_find_gpus patch ideia, just looked easier.

And seatd does not support seat yet, the is code is almost done;
In this case, seat name would be passed with environment var?
 (On weston I use arguments for the seat and the drm lease)

**Comment by garlett** [comment]
But using logind and patching wlroots at [wlr_session_find_gpus](https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/backend/session/session.c?ref_type=heads#L450) and [wlr_seat_create](https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/types/seat/wlr_seat.c?ref_type=heads#L254) could work?

Or patch labwc at [wlr_session_find_gpus where?]() and [wlr_seat_create](https://github.com/labwc/labwc/blob/master/src/seat.c#L519)

Patch wlroots opens more compositor options, but in my case I alredy have to compile labwc from AUR

**Comment by garlett** [comment]
@Consolatis this is a stub for the multiseat patch, can you give a look ? thx
```
diff --git a/backend/session/session.c b/backend/session/session.c
index 6254a734..c017989c 100644
--- a/backend/session/session.c
+++ b/backend/session/session.c
@@ -18,6 +18,7 @@
 #include "util/time.h"
 
 #include <libseat.h>
+#include <dlmclient.h>
 
 #define WAIT_GPU_TIMEOUT 10000 // ms
 
@@ -319,38 +320,46 @@ struct wlr_device *wlr_session_open_file(struct wlr_session *session,
 		return NULL;
 	}
 
+	struct wlr_device *wlr_dev = wlr_fd_to_device(session, fd);
+	if (wlr_dev){
+		return wlr_dev;
+	}
+	libseat_close_device(session->seat_handle, device_id);
+ 	close(fd);
+	return NULL;
+}
+
+struct wlr_device *wlr_fd_to_device(struct wlr_session *session, int fd) { // bottom of wlr_session_open_file 
 	struct wlr_device *dev = malloc(sizeof(*dev));
 	if (!dev) {
 		wlr_log_errno(WLR_ERROR, "Allocation failed");
-		goto error;
+		return NULL;
 	}
 
 	struct stat st;
-	if (fstat(fd, &st) < 0) {
+	if (fstat(fd, &st) < 0 || !S_ISCHR(st.st_mode)) { // !S_ISCHR copied from weston,  need to verify
 		wlr_log_errno(WLR_ERROR, "Stat failed");
-		goto error;
+		free(dev);
+		return NULL;
 	}
 
 	dev->fd = fd;
 	dev->dev = st.st_rdev;
-	dev->device_id = device_id;
+	dev->device_id = st.st_dev; // is correct to chage device_id to st.st_dev ????
 	wl_signal_init(&dev->events.change);
 	wl_signal_init(&dev->events.remove);
 	wl_list_insert(&session->devices, &dev->link);
-
 	return dev;
-
-error:
-	libseat_close_device(session->seat_handle, device_id);
-	free(dev);
-	close(fd);
-	return NULL;
 }
 
 void wlr_session_close_file(struct wlr_session *session,
 		struct wlr_device *dev) {
-	if (libseat_close_device(session->seat_handle, dev->device_id) == -1) {
-		wlr_log_errno(WLR_ERROR, "Failed to close device %d", dev->device_id);
+	if (dev->drm_lease){
+		dlm_release_lease(dev->drm_lease);
+	} else {
+		if (libseat_close_device(session->seat_handle, dev->device_id) == -1) {
+			wlr_log_errno(WLR_ERROR, "Failed to close device %d", dev->device_id);
+		}
 	}
 	close(dev->fd);
 	wl_list_remove(&dev->link);
@@ -444,8 +453,42 @@ static void find_gpus_handle_add(struct wl_listener *listener, void *data) {
 	handler->added = true;
 }
 
+static ssize_t open_kms_lease(struct wlr_session *session, struct wlr_device *dev, const char *drm_lease_name) {
+	wlr_log(WLR_INFO, "Opening KMS device: %s", drm_lease_name);
+
+	int fd;
+	struct dlm_lease *lease = dlm_get_lease(drm_lease_name);
+	if (!lease  || (fd = dlm_lease_fd(lease)) < 0 ) {
+		wlr_log(WLR_DEBUG, "Could not get DRM lease %s", drm_lease_name);
+		dlm_release_lease(lease);
+		return 0;
+	}
+
+	dev = wlr_fd_to_device(session, fd);
+	dev->drm_lease = lease;
+	if (!dev) {
+		wlr_log(WLR_ERROR, "Unable to open %s as KMS device", drm_lease_name);
+		close(fd);
+		dlm_release_lease(lease);
+		return 0;
+	}
+
+	if (!drmIsKMS(dev->fd)) {
+		wlr_log(WLR_DEBUG, "Ignoring '%s': not a KMS device", drm_lease_name);
+		wlr_session_close_file(session, dev);
+		return 0;
+	}
+
+	return 1;
+}
+
 ssize_t wlr_session_find_gpus(struct wlr_session *session,
 		size_t ret_len, struct wlr_device **ret) {
+	const char *lease = getenv("DRM_LEASE");
+	if (lease) {
+		return open_kms_lease(session, ret[0], lease);
+	}
+
 	const char *explicit = getenv("WLR_DRM_DEVICES");
 	if (explicit) {
 		wlr_log(WLR_INFO, "Opening fixed list of KMS devices from WLR_DRM_DEVICES: %s", explicit);
diff --git a/include/wlr/backend/session.h b/include/wlr/backend/session.h
index 3179d1f5..49a4621a 100644
--- a/include/wlr/backend/session.h
+++ b/include/wlr/backend/session.h
@@ -4,6 +4,7 @@
 #include <stdbool.h>
 #include <sys/types.h>
 #include <wayland-server-core.h>
+#include <dlmclient.h>
 
 struct libseat;
 
@@ -20,6 +21,8 @@ struct wlr_device {
 		struct wl_signal change; // struct wlr_device_change_event
 		struct wl_signal remove;
 	} events;
+	
+	struct dlm_lease *drm_lease;
 };
 
 /**
@@ -120,6 +123,12 @@ void wlr_session_destroy(struct wlr_session *session);
 struct wlr_device *wlr_session_open_file(struct wlr_session *session,
 	const char *path);
 
+/*
+ * converts file descriptor to wlr_device
+ * initializes wlr session signals
+ */
+struct wlr_device *wlr_fd_to_device(struct wlr_session *session, int fd);
+
 /*
  * Closes a file previously opened with wlr_session_open_file().
  */
diff --git a/meson.build b/meson.build
index fcc562c8..d0921cde 100644
--- a/meson.build
+++ b/meson.build
@@ -124,6 +124,7 @@ wlr_deps = [
 	math,
 	rt,
 ]
+wlr_deps += dependency('libdlmclient')
 
 subdir('protocol')
 subdir('render')
diff --git a/types/seat/wlr_seat.c b/types/seat/wlr_seat.c
index 2e8e4a84..6671c10c 100644
--- a/types/seat/wlr_seat.c
+++ b/types/seat/wlr_seat.c
@@ -316,7 +316,11 @@ struct wlr_seat *wlr_seat_create(struct wl_display *display, const char *name) {
 		return NULL;
 	}
 	seat->display = display;
-	seat->name = strdup(name);
+	seat->name = getenv("XDG_SEAT");
+	if(!seat->name){
+		seat->name = strdup(name);
+	}
+	
 	wl_list_init(&seat->clients);
 	wl_list_init(&seat->selection_offers);
 	wl_list_init(&seat->drag_offers);
```

**Reply by Consolatis** [reply]
I think having some external application that receives the leased DRM fd and then executes the compositor would be more flexible. Then wlroots would only have to parse some env var and use your `wlr_fd_to_device()`function for the inherited DRM fd and different starter applications could be developed (e.g. one for the automotive thingie and another one for the drm_lease protocol). Mutliseat itself is kind of a different thing though because that should also apply for inputs and not just the DRM fd.

**Reply by garlett** [reply]
kennylevinsen suggested that seatd could create/manage the lease and send the fd to libseat, eliminating the lease manager too.
But until seatd add support multiseat, this could be a temporary patch.
And I do not understand enough of the lease protocol and seatd, to incorporate the lease manager code inside seatd.
I [finished](https://gitlab.freedesktop.org/garlett/wlroots-lease-multiseat/-/compare/master...master?from_project_id=12103) the patch, after test it, you can include what is good in wlroots.
Thanks

**Reply by hezhanbang** [reply]
@garlett Your project [`multiseat`](https://github.com/garlett/multiseat) is great, but I can't get it to work. Can you give me some guidance?

**Reply by garlett** [reply]
> @garlett Your project [`multiseat`](https://github.com/garlett/multiseat) is great, but I can't get it to work. Can you give me some guidance?

Whats happening? ( there is a problem when my wlroots fork is not updated, because I made the script create the patch every time ) ( open an issue, so more people can have help with the same problem)

**Reply by johanmalm** [reply]
Gitlab is being migrated so cannot access wlroots for a few days... is that what you mean??
